File: redis/test-geohash.lua
#!/usr/bin/env luajit

local ffi = require("ffi")

local rgeo = ffi.load("librgeohash.so")

ffi.cdef[[
void geohash(double longitude, double latitude, char* hash);
uint64_t geohash64(double longitude, double latitude);
void geohash_decode(uint64_t bits, double* longitude, double* latitude);
uint64_t geohash_decode64(char* buf);
]]

local function geohash(longitude, latitude)
    local lon = ffi.new("double", longitude)
    local lat = ffi.new("double", latitude)
    local hash = ffi.new("char[?]", 12)
    rgeo.geohash(lon, lat, hash)
    return ffi.string(hash, 11)
end

local function geohash64(longitude, latitude)
    return rgeo.geohash64(longitude, latitude)
end

local function geohash_decode(bits)
    local b = ffi.new("uint64_t", bits)
    local longitude = ffi.new("double[1]", 10.1)
    local latitude = ffi.new("double[1]", 20.2)
    rgeo.geohash_decode(b, longitude, latitude)
    return longitude[0], latitude[0]
end

local function geohash_decode64(hash)
    local h = ffi.new("char[12]", hash)
    return rgeo.geohash_decode64(h)
end

local longitude = -46.691657
local latitude = -23.570018

local hash = geohash(longitude, latitude)
local bits = geohash64(longitude, latitude)

print("geohash...");
print(string.format("longitude: %.8f and latitude: %.8f [%s] = %s\n", longitude, latitude, hash, bits))

longitude, latitude = geohash_decode(bits)
bits = geohash_decode64(hash)

print("geohash (decode)...")
print(string.format("longitude: %.8f and latitude: %.8f [%s] = %s\n", longitude, latitude, hash, bits))

print("testing...")
print(string.format("1.invalid longitude/latitude [%s]", geohash(1, 1111)));

File: redis/test-geohash.pl
use Modern::Perl;
use Data::Dumper;

use Inline C => <<'END_C', LIBS => '-lrgeohash'; 

#include <stdint.h>
#include <string.h>
#include "rgeohash.h"

void geohash_sv(double longitude, double latitude, SV* res) {
    char hash[12];
    geohash(longitude, latitude, hash);
    sv_setpvn(res, hash, strlen(hash));
}

unsigned long geohash64_sv(double longitude, double latitude) {
    return (unsigned long) geohash64(longitude, latitude);
}

void geohash_decode_sv(unsigned long bits, SV* longitude, SV* latitude) {
    double lon, lat;
    geohash_decode(bits, &lon, &lat);
    sv_setnv(longitude, lon);
    sv_setnv(latitude, lat);
}


unsigned long geohash_decode64_sv(char* buf) {
    return (unsigned long) geohash_decode64(buf);
}

END_C

my ($hash, $hash64, $longitude, $latitude, $hash64_sv);

geohash_sv(-46.691657, -23.570018, $hash);

$hash64 = geohash64_sv(-46.691657, -23.570018);

geohash_decode_sv($hash64 + 0, $longitude, $latitude);

$hash64_sv = geohash_decode64_sv($hash);

say(Dumper([$hash, $hash64, $longitude, $latitude, $hash64_sv], 
           [qw(hash hash64 longitude latitude hash64_sv)]));
